solls/LanguageServer.cpp

```diff
@@ -0,0 +1,668 @@
+}
+
+void LanguageServer::operator()(lsp::protocol::DidCloseTextDocumentParams const& _didClose)
+{
+	logMessage("LanguageServer: didClose: " + _didClose.textDocument.uri);
+}
+
+void LanguageServer::validateAll()
+{
+	for (reference_wrapper<lsp::vfs::File const> const& file: m_vfs.files())
+		validate(file.get());
+}
+
+void LanguageServer::validate(lsp::vfs::File const& _file)
+{
+	PublishDiagnosticsList result;
+	validate(_file, result);
+
+	for (lsp::protocol::PublishDiagnosticsParams const& diag: result)
+		notify(diag);
+}
+
+frontend::ReadCallback::Result LanguageServer::readFile(string const& _kind, string const& _path)
+{
+	return m_fileReader->readFile(_kind, _path);
+}
+
+constexpr lsp::protocol::DiagnosticSeverity toDiagnosticSeverity(Error::Type _errorType)
+{
+	switch (_errorType)
+	{
+		case Error::Type::CodeGenerationError:
+		case Error::Type::DeclarationError:
+		case Error::Type::DocstringParsingError:
+		case Error::Type::ParserError:
+		case Error::Type::SyntaxError:
+		case Error::Type::TypeError:
+			return lsp::protocol::DiagnosticSeverity::Error;
+		case Error::Type::Warning:
+			return lsp::protocol::DiagnosticSeverity::Warning;
+	}
+	// Should never be reached.
+	return lsp::protocol::DiagnosticSeverity::Error;
+}
+
+void LanguageServer::compile(lsp::vfs::File const& _file)
+{
+	// TODO: optimize! do not recompile if nothing has changed (file(s) not flagged dirty).
+
+	// always start fresh when compiling
+	m_sourceCodes.clear();
+
+	m_sourceCodes[_file.uri().substr(7)] = _file.contentString();
+
+	m_fileReader = make_unique<FileReader>(m_basePath, m_allowedDirectories);
+
+	m_compilerStack.reset();
+	m_compilerStack = make_unique<CompilerStack>(bind(&FileReader::readFile, ref(*m_fileReader), _1, _2));
+
+	// TODO: configure all compiler flags like in CommandLineInterface (TODO: refactor to share logic!)
+	OptimiserSettings settings = OptimiserSettings::standard(); // TODO: get from config
+	m_compilerStack->setOptimiserSettings(settings);
+	m_compilerStack->setParserErrorRecovery(true);
+	m_compilerStack->setEVMVersion(EVMVersion::constantinople()); // TODO: get from config
+	m_compilerStack->setRevertStringBehaviour(RevertStrings::Default); // TODO get from config
+	m_compilerStack->setSources(m_sourceCodes);
+
+	m_compilerStack->compile();
+}
+
+void LanguageServer::validate(lsp::vfs::File const& _file, PublishDiagnosticsList& _result)
+{
+	compile(_file);
+
+	lsp::protocol::PublishDiagnosticsParams params{};
+	params.uri = _file.uri();
+
+	for (shared_ptr<Error const> const& error: m_compilerStack->errors())
+	{
+		// Don't show this warning. "This is a pre-release compiler version."
+		if (error->errorId().error == 3805)
+			continue;
+
+		auto const message = SourceReferenceExtractor::extract(*error);
+
+		auto const severity = toDiagnosticSeverity(error->type());
+
+		// global warnings don't have positions in the source code - TODO: default them to top of file?
```

Comment:
If there's no source location, don't we just assume it is valid for the entire file?

Reply:
That's right. I could test how this would look like in IDEs when  I specifically the whole file range.